source/develop/release-management/features/virt/tpm-device.md

```diff
@@ -0,0 +1,154 @@
+
+This approach should be feasible with introducing a new Vdsm API call as outlined above but it has the drawback that Engine assistance is needed.
+
+#### Storing the data in domain XML metadata
+
+Using domain XML metadata, it's possible to transfer TPM data in all the directions.  TPM data can be read from the data on VM start, with or without Engine assistance, updates can be stored there atomically and Engine can update the data in monitoring.
+
+Ensuring data consistency read from swtpm is the same as with API calls.
+
+The only problem with using metadata is that it increases domain XML size significantly.  TPM data shouldn't be large although to know the size limits would require further investigation.  But initial swtpm data for a VM takes about 4 KB encoded.  It would "pollute" the domain XML and would have to be transferred each time Engine asks for the domain XML, even when there are no changes to data, which means it would be waste most of the time.
+
+This approach is a bit simpler and more flexible than using API calls but has the data size drawback.
+
+#### Using shared storage
+
+A somewhat similar approach to using domain XML metadata would be to store the data to a shared storage.  Instead of writing to or reading it from the metadata, it would be written to or read from the storage.  The benefit would be avoiding domain XML pollution and unnecessary data transfers.  Additionally, Engine wouldn't have to process the data and maintain it in the database and the data could be updated even when Engine was temporarily unavailable.
+
+However, there are also drawbacks:
+
+- A volume for TPM data must be created, arranged and maintained properly.
+- Vdsm is responsible for arranging atomic data updates, to avoid ending up with partially written data in case storage gets disconnected.
+- Unless direct I/O is used, Vdsm can't be sure the data was actually written.
+
+This approach would be more complicated than using metadata but it would have the advantages mentioned above.
+
+#### Making swtpm to use shared storage directly
+
+The swtpm data location is selected by libvirt automatically but we could probably make a symbolic link to the desired shared location.  This approach would avoid data handling on the Vdsm side and all the updates would be stored immediately.  But it shares some of the drawbacks of using shared storage and there are additional risks:
+
+- Data consistency is at risk when storage gets disconnected during swtpm write.
+- swtpm reacts to write failures [by deleting the data completely](https://github.com/stefanberger/swtpm/blob/609dfd873a31fc0402752c50d1d5cfa58949507f/src/swtpm/swtpm_nvfile.c#L563), which may fail as well, but a question is if the data is ever attempted to write again without further updates.
```

Comment:
deleting on failure is not valid approach. 

The only way that will work is copy-on-write:

- Keep metadata sector pointing to the current data area
- keep 2 data areas

For each write, write to non-current data area. When the write is done
(fsync() succeeded), update the metadata sector to to point to this area.

On failure, the metadata sector still points to previous data area, no change
was done, and there nothing to cleanup.

On success, the metadata sector points to the new data.

If swtmp does not have something like this, there is no point to let it write
to shared storage since it will create inconsistent data.

Reply:
Yes.